src/Microsoft.OData.Core/Json/ODataJsonWriterUtils.cs

```diff
@@ -356,5 +430,183 @@ private static void ODataResourceValueToString(StringBuilder sb, ODataResourceVa
+        /// <param name="instanceAnnotations">Instance annotations for this error.</param>
+        /// <param name="writeInstanceAnnotationsDelegate">Delegate to write the instance annotations.</param>
+        /// <param name="maxInnerErrorDepth">The maximum number of nested inner errors to allow.</param>
+        /// <returns>A task that represents the asynchronous write operation.</returns>
+        private static async Task WriteErrorAsync(
+            IJsonWriterAsync jsonWriter,
+            string code,
+            string message,
+            string target,
+            IEnumerable<ODataErrorDetail> details,
+            ODataInnerError innerError,
+            IEnumerable<ODataInstanceAnnotation> instanceAnnotations,
+            Func<IEnumerable<ODataInstanceAnnotation>, Task> writeInstanceAnnotationsDelegate,
+            int maxInnerErrorDepth)
+        {
+            Debug.Assert(jsonWriter != null, "jsonWriter != null");
+            Debug.Assert(code != null, "code != null");
+            Debug.Assert(message != null, "message != null");
+            Debug.Assert(instanceAnnotations != null, "instanceAnnotations != null");
+
+            // "error": {
+            await jsonWriter.StartObjectScopeAsync().ConfigureAwait(false);
+            await jsonWriter.WriteNameAsync(JsonLightConstants.ODataErrorPropertyName).ConfigureAwait(false);
+            await jsonWriter.StartObjectScopeAsync().ConfigureAwait(false);
+
+            // "code": "<code>"
+            await jsonWriter.WriteNameAsync(JsonConstants.ODataErrorCodeName).ConfigureAwait(false);
+            await jsonWriter.WriteValueAsync(code).ConfigureAwait(false);
+
+            // "message": "<message string>"
+            await jsonWriter.WriteNameAsync(JsonConstants.ODataErrorMessageName).ConfigureAwait(false);
+            await jsonWriter.WriteValueAsync(message).ConfigureAwait(false);
+
+            // For example, "target": "query",
+            if (target != null)
+            {
+                await jsonWriter.WriteNameAsync(JsonConstants.ODataErrorTargetName).ConfigureAwait(false);
+                await jsonWriter.WriteValueAsync(target).ConfigureAwait(false);
+            }
+
+            // Such as, "details": [
+            //  {
+            //   "code": "301",
+            //   "target": "$search",
+            //   "message": "$search query option not supported"
+            //  }]
+            if (details?.Any() == true)
+            {
+                await WriteErrorDetailsAsync(jsonWriter, details, JsonConstants.ODataErrorDetailsName).ConfigureAwait(false);
+            }
+
+            if (innerError != null)
+            {
+                await WriteInnerErrorAsync(jsonWriter, innerError, JsonConstants.ODataErrorInnerErrorName,
+                    /* recursionDepth */ 0, maxInnerErrorDepth).ConfigureAwait(false);
+            }
+
+            if (instanceAnnotations.Any())

```

Comment:
>instanceAnnotations.Any() [](start = 16, length = 25)

probably better if writeInstanceAnnotationsDelegate handles the empty case so that you don't have to use the LINQ Any() method to check for instance annotations before invoking the writeInstanceAnnotationsDelegate (you would need to make the check for writeInstanceAnnotationsDelegate a real check, not just an assertion).

Reply:
@mikepizzo To handle the empty instanceAnnotations in the async invocation of the delegate, we'd have to require the delegate to always be non-null. Which is not a problem since we have [`WriteInstanceAnnotationsForError` method](https://github.com/OData/odata.net/blob/cc9d86f0069be395a4f07aabe5cad0cec7b9a2cd/src/Microsoft.OData.Core/Json/JsonLightInstanceAnnotationWriter.cs#L125) that we can pass as the delegate - defined in `JsonLightAnnotationWriter`.
If we're on the same page on that, I'll change the logic such that we throw an exception if the delegate is `null`, firstly from the `internal WriteErrorAsync` method - so we don't bother to call the `private WriteErrorAsync`. I'll add the same check in the `private` method just in case it was triggered from another method